store/v2/root/store.go

```diff
@@ -250,21 +249,39 @@ func (s *Store) LoadVersionAndUpgrade(version uint64, upgrades *corestore.StoreU
 		return err
 	}
 
-	// if upgrades is not nil, we need to add the pruned KVStores to the pruning manager
+	// if upgrades is not nil, we need to prune the old store keys from SS and SC
 	if upgrades != nil {
+		removedStoreKeys := make([]string, 0)
+		removedStoreKeys = append(removedStoreKeys, upgrades.Deleted...)
+		for _, renamedStore := range upgrades.Renamed {
+			removedStoreKeys = append(removedStoreKeys, renamedStore.OldKey)
+		}
+		// if the state commitment implements the KVStoreGetter interface, prune the
+		// old KVStores
 		kvStoreGetter, ok := s.stateCommitment.(store.KVStoreGetter)
 		if !ok {
+			s.logger.Warn("state commitment does not implement KVStoreGetter interface")
+		} else {
 			removedKVStores := make([]corestore.KVStoreWithBatch, 0)
-			for _, storeKey := range upgrades.Deleted {
+			for _, storeKey := range removedStoreKeys {
 				removedKVStores = append(removedKVStores, kvStoreGetter.GetKVStoreWithBatch(storeKey))
 			}
-			for _, renamedStore := range upgrades.Renamed {
-				removedKVStores = append(removedKVStores, kvStoreGetter.GetKVStoreWithBatch(renamedStore.OldKey))
-			}
 			if err := s.pruningManager.PruneKVStores(removedKVStores); err != nil {
 				return fmt.Errorf("failed to set pruned KVStores: %w", err)
 			}
 		}
+		// if the state storage implements the UpgradableDatabase interface, prune the
+		// old store keys
+		upgradableDatabase, ok := s.stateStorage.(store.UpgradableDatabase)
+		if !ok {
+			s.logger.Warn("state storage does not implement UpgradableDatabase interface")
+		} else {
+			for _, storeKey := range removedStoreKeys {
+				if err := upgradableDatabase.PruneStoreKey([]byte(storeKey)); err != nil {
```

Comment:
Would it make sense to pass all `removedStoreKeys` at once to let the DB implementation decide on the best batch size and strategy?